File: easyboard-backend_blank/easyboard-aggregate/src/main/java/io/naradrama/easyboard/aggregate/posting/domain/entity/Posting.java
/*
 COPYRIGHT (c) NEXTREE Inc. 2014
 This software is the proprietary of NEXTREE Inc.
 @since 2014. 6. 10.
*/
package io.naradrama.easyboard.aggregate.posting.domain.entity;

import io.naradrama.easyboard.aggregate.posting.domain.entity.sdo.PostingCdo;
import io.naradrama.prologue.domain.NameValue;
import io.naradrama.prologue.domain.NameValueList;
import io.naradrama.prologue.domain.ddd.DomainAggregate;
import io.naradrama.prologue.domain.ddd.DomainEntity;
import io.naradrama.prologue.util.json.JsonUtil;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.beans.BeanUtils;

@Getter
@Setter
@NoArgsConstructor
public class Posting extends DomainEntity implements DomainAggregate {
    //
    private String writerId;
    private String writerName;
    private String title;
    private String content;
    private String base64AttachedImage;
    private String boardId;
    private boolean important;
    private boolean deleted;
    private long time;

    public Posting(String id) {
        //
        super(id);
    }

    public Posting(PostingCdo postingCdo) {
        super();
        BeanUtils.copyProperties(postingCdo, this);
        this.important = false;
        this.deleted = false;
        this.time = System.currentTimeMillis();
    }

    public void markDeleted() {
        //
        this.deleted = true;
    }

    public void modifyValues(NameValueList nameValues) {
        //
        for (NameValue nameValue : nameValues.list()) {
            String value = nameValue.getValue();
            switch (nameValue.getName()) {
                case "title":
                    this.title = value;
                    break;
                case "content":
                    this.content = value;
                    break;
                case "base64AttachedImage":
                    this.base64AttachedImage = value;
                    break;
                case "important":
                    this.important = Boolean.parseBoolean(value);
                    break;

                default:
                    throw new IllegalArgumentException("Update not allowed: " + nameValue);
            }
        }
    }

    @Override
    public String toString() {
        //
        return toJson();
    }

    public static Posting fromJson(String json) {
        //
        return JsonUtil.fromJson(json, Posting.class);
    }

    public static Posting sample() {
        //
        return new Posting(PostingCdo.sample());
    }

    public static void main(String[] args) {
        //
        System.out.println(sample().toPrettyJson());
    }
}

File: easyboard-frontend-blank/src/comp/api/posting/api-model/Posting.ts
import {
  NameValueList,
  fromDomain,
  DomainEntity,
  makeExtendedObservable,
} from '@nara.drama/prologue';
import moment from 'moment';


@fromDomain
class Posting extends DomainEntity {
  writerId: string;
  writerName: string;
  title: string;
  content: string;
  base64AttachedImage: string;
  boardId: string;
  important: boolean;
  deleted: boolean;
  time: number;

  /* for ui */
  expanded: boolean = false;
  editing: boolean = false;

  constructor(
    writerId: string,
    writerName: string,
    title: string,
    content: string,
    base64AttachedImage: string,
    boardId: string,
    important: boolean,
    deleted: boolean,
    time: number,
  ) {
    super();
    this.writerId = writerId;
    this.writerName = writerName;
    this.title = title;
    this.content = content;
    this.base64AttachedImage = base64AttachedImage;
    this.boardId = boardId;
    this.important = important;
    this.deleted = deleted;
    this.time = time;

    makeExtendedObservable(this);
  }

  static fromDomain(domain: Posting): Posting {
    const posting = new Posting(
      domain.writerId,
      domain.writerName,
      domain.title,
      domain.content,
      domain.base64AttachedImage,
      domain.boardId,
      domain.important,
      domain.deleted,
      domain.time,
    );

    posting.setDomainEntity(domain);
    return posting;
  }

  static fromDomains(domains: Posting[]): Posting[] {
    return domains.map(domain => this.fromDomain(domain));
  }

  static asNameValues(posting: Posting) {
    //
    return NameValueList.fromModel(Posting, posting, {
      title: String,
      content: String,
      base64AttachedImage: String,
      important: String,
      // embeddedSubPostings: JSON,
    });
  }

  get displayTime() {
    //
    return moment(this.time).format('YYYY-MM-DD');
  }
}

export default Posting;
